### PR TITLE
chore: email treasury report zip file

### PR DESCRIPTION
**Pull request by vshia**
Need help on the Terraform setup for AWS SES

Task: https://github.com/usdigitalresponse/cpf-reporter/issues/321

### CODE REVIEW

File: python/pyproject.toml [hunk 1]
```diff
@@ -12,6 +12,8 @@ pydantic = "^2.6.4"
 aws-lambda-typing = "^2.19.0"
 boto3 = "^1.34.70"
 structlog = "^24.1.0"
+html2text = "^2024.2.26"
```
Comment (vshia) [inline python/pyproject.toml:15]: will remove

File: python/src/functions/generate_presigned_url_and_send_email.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+Your treasury report can be downloaded here: {url}.
+"""
+
+class ProjectLambdaPayload(BaseModel):
+    organization: OrganizationObj
+    user: UserObj
+
+
+@reset_contextvars
+def handle(event: ProjectLambdaPayload, context: Context):
```
Comment (as1729) [inline python/src/functions/generate_presigned_url_and_send_email.py:45]: @vshia just a note that the naming of the payload and some of the logs on this function will need to be updated. I assume this is not yet changed due to the draft nature of the PR.

**Review by as1729 — CHANGES_REQUESTED** [review]
@vshia I have requested a few changes on the existing code. I will leave comments about Amazon SES after testing it out locally first by EOD wednesday.

File: terraform/email.tf [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+resource "aws_ses_configuration_set" "ses_config" {
+  name = "config_ses"
+  reputation_metrics_enabled = true
+}
+
+resource "aws_ses_domain_identity" "domain_identity" {
+  domain = var.domain_name
+}
+
+resource "aws_ses_domain_dkim" "dkim_identity" {
```
Comment (as1729) [inline terraform/email.tf:11]: resource not necessary already configured in the USDR account bootstrap repo not accessible to volunteers

Confirming that this is already setup in AWS account for CPF
image

File: terraform/email.tf [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  domain = aws_ses_domain_identity.domain_identity.domain
+}
+
+resource "aws_route53_record" "amazonses_dkim_record" {
```
Comment (as1729) [inline terraform/email.tf:15]: see above- resource is pre-configured

File: terraform/email.tf [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  count   = 3
+  zone_id = aws_route53_zone.route_53_zone.zone_id
+  name    = "${aws_ses_domain_dkim.dkim_identity.dkim_tokens[count.index]}._domainkey.${aws_ses_domain_identity.domain_identity.domain}"
+  type    = "CNAME"
+  ttl     = "300"
+  records = ["${aws_ses_domain_dkim.dkim_identity.dkim_tokens[count.index]}.dkim.amazonses.com"]
+}
+
+resource "aws_ses_domain_identity_verification" "domain_identity_verification" {
```
Comment (as1729) [inline terraform/email.tf:24]: see above- resource is pre-configured

File: terraform/email.tf [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  domain = aws_ses_domain_identity.domain_identity.id
+  depends_on = [aws_route53_record.amazonses_dkim_record]
+}
+
+resource "aws_ses_email_identity" "email_identity" {
+  email = "contact@mahiratechnology.com"
+}
+
+resource "aws_route53_zone" "route_53_zone" {
```
Comment (as1729) [inline terraform/email.tf:33]: see above- resource is pre-configured

File: terraform/variables.tf [hunk 7]
```diff
@@ -315,3 +315,36 @@ variable "postgres_ca_cert_identifier" {
...
+  type        = bool
+  default     = false
+}
+
+variable "domain_name" {
+  type = string
+  default = "yourdomain.com"
+}
+
+variable "ses_sandbox_mode_email_recipients" {
```
Comment (as1729) [inline terraform/variables.tf:332]: can be ignored for CPF and left out of the PR

**Review by as1729 — COMMENTED** [review]
@vshia I added a comment indicating the removal of a few resources. Later today I will provide you with a code-snippet of what resources or in our case data that we need to add here.

**Comment by as1729** [comment]
@vshia this PR contains the changes I would make to this PR in order to get the SES piece to work: https://github.com/usdigitalresponse/cpf-reporter/pull/405

File: terraform/functions.tf [hunk 8]
```diff
@@ -554,3 +554,67 @@ module "lambda_function-cpfValidation" {
     }
   }
 }
+
+module "lambda_function-email-presigned-url" {
```
[the remark this answers is not in the document]
Comment (as1729) [reply terraform/functions.tf:558]: done in PR #405

File: terraform/functions.tf [hunk 9]
```diff
@@ -554,3 +554,67 @@ module "lambda_function-cpfValidation" {
...
+  layers        = local.lambda_py_layer_arns
+  timeout       = 60 # 1 minute, in seconds
+  memory_size   = 512
+  environment_variables = merge(local.lambda_default_environment_variables, {
+    DD_LAMBDA_HANDLER = "src.functions.generate_presigned_url_and_send_email"
+    DD_LOGS_INJECTION = "true"
+  })
+
+  // Triggers
+  allowed_triggers = {
```
[the remark this answers is not in the document]
Comment (as1729) [reply terraform/functions.tf:614]: done in PR #405

File: terraform/email.tf (hunk 3 above)
Comment (as1729) [reply terraform/email.tf:11]: Done in PR #405

File: terraform/email.tf (hunk 4 above)
Comment (as1729) [reply terraform/email.tf:15]: Done in PR #405

File: python/src/functions/generate_presigned_url_and_send_email.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    try:
+        payload = ProjectLambdaPayload.model_validate(event)
+    except Exception:
+        logger.exception("Exception parsing Project event payload")
+        return {"statusCode": 400, "body": "Bad Request"}
+
+    try:
+        process_event(payload, logger)
+    except Exception:
+        logger.exception("Exception processing Subrecipient file generation event")
```
Comment (as1729) [inline python/src/functions/generate_presigned_url_and_send_email.py:58]:
```suggestion
        logger.exception("Exception processing sending treasury report email")
```

File: python/src/functions/generate_presigned_url_and_send_email.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+treasury_email_html = """
+Your treasury report can be downloaded <a href={url}>here</a>.
+"""
+
+treasury_email_text = """
+Hello,
+Your treasury report can be downloaded here: {url}.
+"""
+
+class ProjectLambdaPayload(BaseModel):
```
Comment (as1729) [inline python/src/functions/generate_presigned_url_and_send_email.py:27]:
```suggestion
class SendTreasuryEmailLambdaPayload(BaseModel):
```

File: python/src/lib/s3_helper.py [hunk 12]
```diff
@@ -56,3 +56,37 @@ def upload_generated_file_to_s3(
...
+        expiration_time: int = 60 * 60,  #  1 hour
+):
+    logger = get_logger()
+    try:
+        response = s3_client.head_object(
+            bucket = bucket,
+            key = key,
+        )
+    except ClientError as e:
+        if logger:
```
Comment (as1729) [inline python/src/lib/s3_helper.py:74]: This `if logger:` check shouldn't be necessary. Was this added due to issues with testing the functionality?

File: python/src/lib/s3_helper.py [hunk 13]
```diff
@@ -56,3 +56,37 @@ def upload_generated_file_to_s3(
...
+        response =  s3_client.generate_presigned_url(
+            "get_object",
+            Params={
+                "Bucket": bucket,
+                "Key": key,
+            },
+            ExpiresIn=expiration_time
+        )
+    except ClientError as e:
+        if logger:
```
Comment (as1729) [inline python/src/lib/s3_helper.py:88]: Same as above- we shouldn't need to check whether `logger` exists.

File: python/src/lib/s3_helper.py [hunk 14]
```diff
@@ -56,3 +56,37 @@ def upload_generated_file_to_s3(
...
 
     logger.info("successfully uploaded file to s3")
+
+
+def get_presigned_url(
+        s3_client: S3Client,
+        bucket: str,
+        key: str,
+        expiration_time: int = 60 * 60,  #  1 hour
+):
```
Comment (as1729) [inline python/src/lib/s3_helper.py:66]:
```suggestion
) -> Optional[str]:
```

File: python/src/lib/email.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
+import os
+import boto3
+from botocore.exceptions import ClientError
+
+CHARSET = "UTF-8"
+
+
+def send_email(dest_email: str, email_html: str, email_text: str, subject: str, logger):
```
Comment (as1729) [inline python/src/lib/email.py:8]:
```suggestion
def send_email(dest_email: str, email_html: str, email_text: str, subject: str, logger) -> bool:
```

File: python/src/functions/generate_presigned_url_and_send_email.py [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+                email_text = treasury_email_text.format(url=presigned_url)
+                subject = "USDR CPF Treasury Report"
+                return email_html, email_text, subject
+    except Exception as e:
+        logger = get_logger()
+        logger.error(f"Failed to generate treasury email: {e}")
+    return None, None, None
+
+
+def process_event(payload: ProjectLambdaPayload, logger):
```
Comment (as1729) [inline python/src/functions/generate_presigned_url_and_send_email.py:97]:
```suggestion
def process_event(payload: ProjectLambdaPayload, logger:  structlog.stdlib.BoundLogger) -> bool:
```

File: python/src/functions/generate_presigned_url_and_send_email.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+    try:
+        process_event(payload, logger)
+    except Exception:
+        logger.exception("Exception processing Subrecipient file generation event")
+        return {"statusCode": 500, "body": "Internal Server Error"}
+
+    return {"statusCode": 200, "body": "Success"}
+
+
+def generate_email(user: UserObj, presigned_url: str = ""):
```
Comment (as1729) [inline python/src/functions/generate_presigned_url_and_send_email.py:64]:
```suggestion
def generate_email(user: UserObj, presigned_url: str = "", logger: structlog.stdlib.BoundLogger) -> Tuple[Optional[str], Optional[str], Optional[str]]:
```

File: python/src/functions/generate_presigned_url_and_send_email.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+        context: Lambda context
+    """
+    structlog.contextvars.bind_contextvars(lambda_event={"step_function": event})
+    logger = get_logger()
+    logger.info("received new invocation event from step function")
+
+    try:
+        payload = ProjectLambdaPayload.model_validate(event)
+    except Exception:
+        logger.exception("Exception parsing Project event payload")
```
Comment (as1729) [inline python/src/functions/generate_presigned_url_and_send_email.py:52]:
```suggestion
        logger.exception("Exception parsing Send Treasury Email event payload")
```

File: python/src/functions/generate_presigned_url_and_send_email.py [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+    organization = payload.organization
+    
+    presigned_url = get_presigned_url(
+        s3_client = s3_client,
+        bucket = os.getenv("REPORTING_DATA_BUCKET_NAME"),
+        key = f"treasuryreports/{organization.id}/{organization.preferences.current_reporting_period_id}/report.zip",
+        expiration_time = 60 * 60,  # 1 hour
+    )
+    if presigned_url is None:
+        raise
```
Comment (as1729) [inline python/src/functions/generate_presigned_url_and_send_email.py:118]:
```suggestion
        raise Exception('Failed to generate signed-URL or file not found')
```

File: python/src/functions/generate_presigned_url_and_send_email.py [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+                    "email_body": email_body,
+                },
+                partials_dict = {
+                    "email_body": email_body,
+                })
+                email_text = treasury_email_text.format(url=presigned_url)
+                subject = "USDR CPF Treasury Report"
+                return email_html, email_text, subject
+    except Exception as e:
+        logger = get_logger()
```
Comment (as1729) [inline python/src/functions/generate_presigned_url_and_send_email.py:92]:
```suggestion
```

**Review by as1729 — COMMENTED** [review]
Some additional changes @vshia. Mainly related to typing, naming, and error handling. Ideally if we can have custom error message types, that would be the best so we are can more easily identify and triage errors when we see them in datadog, but for now having descriptive error messages will do.

File: python/src/lib/s3_helper.py (hunk 12 above)
Comment (vshia) [reply python/src/lib/s3_helper.py:74]: Yes, I'll remove it. I don't think it's necessary anymore.

**Review by as1729 — APPROVED** [review]

**Comment by as1729** [comment]
Thanks @vshia looks great!

**Pull request merged**